server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetadataUpdater.java

```diff
@@ -129,9 +137,16 @@ public Metadata applyChanges(Metadata oldMetadata, RoutingTable newRoutingTable,
                     updates,
                     minClusterTransportVersion
                 );
-                updatedIndexMetadata = updates.increaseTerm
-                    ? updatedIndexMetadata.withIncrementedPrimaryTerm(shardId.id())
-                    : updatedIndexMetadata;
+
+                // TODO: primaries? or replicas too?
```

Comment:
I think this just needs to consider the nodes that primaries are allocated to, as replicas are re-created from the primary.

Reply:
I am not sure this is going to hold in the existing architecture:
* the node that holds replica is upgraded first
* the shards accepts the write
* replica is refreshed or merged

Reply:
Right - so we need to consider replicas separately? This means we need to store an indexversion for each shard replica, not just by the `ShardId`?